### DISCUSSION TITLE
How to trigger the `Upload` Component using an external button?

### DISCUSSION THREAD

**Discussion by ghost** [ant-design/ant-design#41902] [Q&A]
I'm trying to use an upload component but have the trigger: a button, not be the Upload component itself. I only want it as a preview of the image I'm uploading. How would I trigger the Upload component by click the external button? 

Example:

```
<Upload>
   { . . . }
</Upload>

<Button onClick={triggerUpload}>Upload</Button>
```

**Comment by jalog** [comment]
beforeUpload 返回 false 后，手动上传文件。 请查看 [https://ant-design.antgroup.com/components/upload-cn#components-upload-demo-upload-manually](https://ant-design.antgroup.com/components/upload-cn#components-upload-demo-upload-manually)

**Comment by bombillazo** [comment]
`return false` inside the `beforeUpload` prop callback function.

**Comment by linxianxi** [comment]
```
document.querySelector('.ant-upload input').click()
```

**Comment by tlvietdanh** [comment]
```
triggerUpload = () => {
  if(!uploadRef.current) return
  
  uploadRef.current.click()
}

<Upload>
   <div ref={uploadRef}>
     { . . . }
   </div>
</Upload>

<Button onClick={triggerUpload}>Upload</Button>
```

You can try this

**Comment by KurtLopezSE** [comment]
```
const uploadRef = useRef(null);
  const handleButtonClick = (e) => {
    uploadRef.current?.upload?.uploader?.onClick(e);
  };
  return (
    <div className="App">
      <Upload ref={uploadRef} />
      <Button onClick={handleButtonClick}>test</Button>
    </div>
  );
```

Please take note to pass the event to the onClick

Working sandbox
https://codesandbox.io/p/sandbox/antd-upload-3g8dc5

**Reply by BogdanGorelkin** [reply]
Not bad, however if you are developing under typescript, there will be an error
```Property 'uploader' is private and only accessible within class 'Upload'.ts(2341)```.

![Screenshot 2024-07-18 at 16 43 29](https://github.com/user-attachments/assets/43532e2a-81c9-4a37-9038-40218c79ce72)

To make it work I would suggest overriding it's interface like this:

```ts
import { Upload } from 'antd'
import RcUpload from 'rc-upload'
import { UploadRef } from 'antd/es/upload/Upload'
interface ExtendedUploadRef<T = any> extends Omit<UploadRef<T>, 'upload'> {
  upload: Omit<RcUpload, 'uploader'> & {
    uploader: any
  }
}

export default function YourComponent(){
  const uploadRef = useRef<ExtendedUploadRef>(null)
  // some other logic of your component
  return (
    <>
          <Upload
             ref={uploadRef}
             openFileDialogOnClick={false}
             {...restOfYourProps}
             >
                  <button> uploader </button>
             </Upload>
             {/*some other components*/}
    </>
  )
}
```
